portfolio/src/main/webapp/script.js

```diff
@@ -497,14 +455,160 @@ function locationToArray(){
     return locArray;
 }
 
-/** Hardcoded function to get the waypoint for a given grid */
-function getWaypointFromGrid() {
-    fetch('/grids?requestRow=0&requestCol=0')
-    .then(response => console.log(response.json()))
-    .then((waypoint) => {
-        console.log('Row 0 Col 0, waypoint: ' + waypoint);
+/** Receives an array of routes, finds the safest one using a rating system */
+function pickSafestRoute(routesFromSafeNeighborGrids){
+  var counter = 0;
+  var routeIndex = 0;
+  var lessCrimesInRoute = 1000;
+
+  for (var r = 0; r < routesFromSafeNeighborGrids.routes.length; r++) {
+    var route = result.routes[r];
+
+    for (var j = 0; j < routesFromSafeNeighborGrids.legs[0].steps.length; j++) {
+      var routeArray = new google.maps.Polyline({
+        path: [
+            new google.maps.LatLng(routesFromSafeNeighborGrids.legs[0].steps[j].start_location.lat(),
+              routesFromSafeNeighborGrids.legs[0].steps[j].start_location.lng()),
+            new google.maps.LatLng(routesFromSafeNeighborGrids.legs[0].steps[j].end_location.lat(), 
+              routesFromSafeNeighborGrids.legs[0].steps[j].end_location.lng())
+        ]
+      });
+      
+      fetch('/markers').then(response => response.json()).then((markers) => {
+        markers.forEach((marker) => {
+          var myPosition = new google.maps.LatLng(marker.lat, marker.lng);
+          
+          if (google.maps.geometry.poly.isLocationOnEdge(myPosition, routeArray, 0.0064)) {
+            counter += rateCrime(marker.crimeType);
+            console.log("Rating" + counter);
+            console.log(marker.lat);
+            console.log(marker.lng);
+          }
+        });
+      });
+    }
+    
+    console.log("Rating" + counter);
+    
+    if (counter < lessCrimesInRoute) {
+      lessCrimesInRoute = counter;
+      routeIndex = r;
+    }
+    
+    counter = 0;
+  }
+
+  console.log("Chosen route index: " + routeIndex);
+
+  new google.maps.DirectionsRenderer({
+    map: map,
+    directions: result,
+    routeIndex: routeIndex
+  });    
+}
+
+function getWaypointForGrid(grid) {
+  fetch('/grids?requestRow=' + grid.row + '&requestCol=' + grid.col)
```

Comment:
fetch is an asynchronous call that doesn't block. you need to await here to block on the result. Otherwise it returns nothing immediately. I believe the function getWaypointForGrid also needs to have the `async` keyword

response = await fetch(...);`
waypoint = await response.json();
return {
      location: new google.maps.LatLng(waypoint.lat, waypoint.lng),
      stopover: true
    }

Reply:
Oh, I used hard-coded values to test it and it worked but I will change it to avoid future problems with that, thank you!